File: exp-01/q2-max-min-recursion/solution.cpp
#include <iostream>
#include <climits>
using namespace std;

void findMinMax(int arr[], int n, int &min, int &max)
{
    if (n == 1)
    {
        min = max = arr[0];
        return;
    }
    findMinMax(arr, n - 1, min, max);
    if (arr[n - 1] < min)
        min = arr[n - 1];
    if (arr[n - 1] > max)
        max = arr[n - 1];
}

int main()
{
    int n;
    cin >> n;
    int arr[n];
    for (int i = 0; i < n; i++)
        cin >> arr[i];
    int min = INT_MAX, max = INT_MIN;
    findMinMax(arr, n, min, max);
    cout << "Min: " << min << endl;
    cout << "Max: " << max << endl;
    return 0;
}

File: exp-01/q2-max-min-recursion/solution.go
package main

import "fmt"

func findMinMax(arr []int, n int) (int, int) {
	if n == 1 {
		return arr[0], arr[0]
	}
	curMin, curMax := findMinMax(arr, n-1)
	if arr[n-1] < curMin {
		curMin = arr[n-1]
	}
	if arr[n-1] > curMax {
		curMax = arr[n-1]
	}
	return curMin, curMax
}

func main() {
	var n int
	fmt.Scan(&n)
	arr := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&arr[i])
	}
	minVal, maxVal := findMinMax(arr, n)
	fmt.Printf("Min: %d\n", minVal)
	fmt.Printf("Max: %d\n", maxVal)
}